File: apps/web/lib/directory.ts
import { eq } from "drizzle-orm";
import { q } from "./db/query";
import { membersOf, orgsOf, roleOf } from "./orgs";
import { grantableScopes, grantsOf, ROLE_INFO, type Scope } from "./permissions";
import { appsOf, projectsOf } from "./projects";
import { teamsOf } from "./teams";

export async function organizationsOf(userId: string) {
  const orgs = await orgsOf(userId);
  return Promise.all(
    orgs.map(async (o) => {
      const role = await roleOf(userId, o.id);
      return { id: o.id, name: o.name, slug: o.slug, role, role_label: role ? ROLE_INFO[role]?.label ?? role : null, permissions: grantsOf(role), grantable_scopes: grantableScopes(role) };
    }),
  );
}

export async function projectsDirectory(orgId: string, projectId: string | null, appId: string | null) {
  const projects = await projectsOf(orgId);
  const visible = projectId ? projects.filter((p) => p.id === projectId) : projects;
  return Promise.all(
    visible.map(async (p) => {
      const apps = (await appsOf(p.id)).filter((a) => !appId || a.id === appId);
      return { id: p.id, name: p.name, slug: p.slug, description: p.description, team: p.teamId ? { id: p.teamId, name: p.teamName } : null, apps: apps.map((a) => ({ id: a.id, name: a.name, registry_id: a.registryId, source_host_id: a.sourceHostId, last_synced_at: a.lastSyncedAt })) };
    }),
  );
}

export async function teamsDirectory(orgId: string) {
  const { db, t } = await q();
  const teams = await teamsOf(orgId);
  return Promise.all(
    teams.map(async (team) => {
      const members = await db.select({ userId: t.teamMember.userId, name: t.user.name, email: t.user.email }).from(t.teamMember).innerJoin(t.user, eq(t.teamMember.userId, t.user.id)).where(eq(t.teamMember.teamId, team.id));
      const projects = await db.select({ id: t.project.id, name: t.project.name }).from(t.project).where(eq(t.project.teamId, team.id));
      return { id: team.id, name: team.name, slug: team.slug, description: team.description, members, projects };
    }),
  );
}

export async function membersDirectory(orgId: string) {
  return (await membersOf(orgId)).map((m) => ({ user_id: m.userId, name: m.name, email: m.email, role: m.role, role_label: ROLE_INFO[m.role as keyof typeof ROLE_INFO]?.label ?? m.role }));
}

export type Effective = { role: string | null; permissions: Record<string, boolean>; scope: Scope[] | null };

File: action_platform/api/schemas/directory.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class Named(BaseModel):
    id: str
    name: str


class OrganizationRow(BaseModel):
    id: str
    name: str
    slug: str
    role: Optional[str] = None
    role_label: Optional[str] = None
    permissions: dict[str, bool]
    grantable_scopes: list[str]


class AppRef(BaseModel):
    id: str
    name: str
    registry_id: str


class Me(BaseModel):
    user: dict
    organization: Optional[dict] = None
    organizations: Optional[list[dict]] = None
    role: Optional[str] = None
    role_label: Optional[str] = None
    scope: Optional[list[str]] = None
    permissions: dict[str, bool]
    token: Optional[str] = None
    project: Optional[Named] = None
    app: Optional[AppRef] = None


class AppInProject(BaseModel):
    id: str
    name: str
    registry_id: str
    source_host_id: Optional[str] = None
    last_synced_at: Optional[datetime] = None


class ProjectRow(BaseModel):
    id: str
    name: str
    slug: str
    description: Optional[str] = None
    team: Optional[Named] = None
    apps: list[AppInProject]
    organization: Optional[Named] = None


class TeamMemberRow(BaseModel):
    userId: str
    name: str
    email: str


class TeamRow(BaseModel):
    id: str
    name: str
    slug: str
    description: Optional[str] = None
    members: list[TeamMemberRow]
    projects: list[Named]


class MemberRow(BaseModel):
    user_id: str
    name: str
    email: str
    role: str
    role_label: str


class CreateProjectRequest(BaseModel):
    name: str
    description: Optional[str] = ""


class CreateTeamRequest(BaseModel):
    name: str
    description: Optional[str] = ""


class TeamMemberRequest(BaseModel):
    team_id: str
    user_id: str


class ProjectTeamRequest(BaseModel):
    project_id: str
    team_id: Optional[str] = None


class MemberRoleRequest(BaseModel):
    user_id: str
    role: str


class Created(BaseModel):
    id: str
    name: str
    slug: str


class Ok(BaseModel):
    ok: bool = True


class IssueRequest(BaseModel):
    scope: str
    name: Optional[str] = None


class Issued(BaseModel):
    token: str
    token_type: str = "Bearer"
    id: str
    scope: str
    expires_at: datetime
    organization: Optional[dict] = None
    organizations: Optional[list[dict]] = None
    project: Optional[Named] = None
    app: Optional[AppRef] = None


class RoleInfo(BaseModel):
    id: str
    label: str
    description: str
    permissions: list[str]
    grantable_scopes: list[str]


class PermissionInfo(BaseModel):
    id: str
    description: str


class ScopeInfo(BaseModel):
    id: str
    label: str
    description: str
    permissions: list[str]


class AccessCatalog(BaseModel):
    roles: list[RoleInfo]
    permissions: list[PermissionInfo]
    scopes: list[ScopeInfo]
    default_scopes: list[str]
